wordpress-mobile/WordPress-Android
# Auto-upload/save all posts/pages with local changes

Fixes #10174

This PR changes more than I initially planned. I think it'd be worth considering doing a pair/triple-review with me so I can explain the changes. Please let me know if that's what you prefer to do;).


Main changes
1. PostModel.changesConfirmedContentHashcode is a field which we use to mark that the current set of changes has been explicitly confirmed by the user. Whenever the user explicitly confirms the chagnes - clicks on Publish/Update/Schedule/Submit/Save/Sync/... button, we calculate a hashcode of the content of the post and save its value into the `changesConfirmedContentHashcode` field. When we are about to upload the post we calculate the hashcode again and if it matches the value stored in `changesConfirmedContentHashcode` we are sure the post hasn't been changed since the user confirmed the changes and we can safely upload it.

2. PostUploadHandler is the place where we decide whether the post will be uploaded, uploaded as draft or remote auto saved.

- **Uploaded** are posts with changes which have been explicitly confirmed by the user -> the user clicked on Publish/Update/Schedule/Submit/Save/Sync... buttons.

- **Uploaded as drafts** are local drafts (PostModel.isLocalDraft == true) with changes which haven't been confirmed by the user. We can't remote auto save them as they don't exist in the remote.

- **Remote-auto-saved** are all other changes. This feature isn't supported on self-hosted sites, therefore we don't sync the changes with the remote on self-hosted sites unless the user explicitly confirms them.

3. UploadStarter now automatically syncs all changes with the remote - PostUploadHandler decides whether the changes should be uploaded, uploaded as draft or remote auto saved. This PR also adds several safe-guards so we don't automatically upload something we didn't want to upload.



To test:
This PR touches 
- manual publishing/saving/updating/...
- automatic publishing/saving/updating/...
- remote previews
- both posts and pages

We decided we'll test these changes when we implement #10207 

Update release notes:

- [ x ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

Notes:
- **This PR should be reviewed by more than one person since the changes are touching auto-uploading, remote-auto-saving and even uploading initiated by the user.**
- I haven't tested Pages at all yet
- This PR targets `master-auto-upload` which was created from `master-remote-preview`.
